File: commands/race.js
// File: commands/race.js

const { EmbedBuilder } = require('discord.js');
const raceManager = require('../utils/raceManager');
const betManager = require('../utils/betManager');

module.exports = {
  name: 'race',
  description: 'Bắt đầu cuộc đua ngựa',
  async execute(message, args, client, fromPrerace = false) {
    try {
        // --- 1. KIỂM TRA ĐIỀU KIỆN ---
        if (raceManager.isPreraceInProgress() && !fromPrerace) {
            return message.reply('⏳ **Đang đếm ngược!** Vui lòng đợi hết thời gian chờ.');
        }

        if (raceManager.isRaceInProgress()) {
          return message.reply('Cuộc đua đang diễn ra. Vui lòng đợi kết thúc!');
        }
        
        const bets = betManager.getAllBets();
        if (bets.size === 0) {
          if (fromPrerace) raceManager.setPreraceStatus(false);
          return message.reply('Chưa có ai đặt cược. Cuộc đua bị hủy! Hãy dùng `!prerace` lại.');
        }
        
        if (Object.keys(raceManager.getCurrentNames()).length === 0) {
            raceManager.generateRaceNames();
        }
        
        // --- 2. BẮT ĐẦU ĐUA ---
        raceManager.setRaceStatus(true);
        if (fromPrerace) raceManager.setPreraceStatus(false);
        
        const startEmbed = new EmbedBuilder()
          .setTitle('🏇 CUỘC ĐUA BẮT ĐẦU!')
          .setColor('#0099ff')
          .setDescription('Các tay đua kiệt xuất đã rời vạch xuất phát!')
          .setTimestamp();
        
        const raceMessage = await message.channel.send({ embeds: [startEmbed] });
        
        const trackLength = 15; 
        let positions = Array(raceManager.HORSE_COUNT).fill(0); 
        let raceFinished = false;
        
        // --- 3. VÒNG LẶP CHẠY ĐUA ---
        while (!raceFinished) {
          const prevPositions = [...positions];
          const newPositions = raceManager.simulateRaceStep(positions, trackLength);
          for (let i = 0; i < positions.length; i++) positions[i] = newPositions[i];
          
          const statusMessage = raceManager.createRaceStatusMessage(positions, trackLength);
          await raceMessage.edit({ content: statusMessage, embeds: [] });
          
          const leadingMessage = raceManager.createLeadingHorseMessage(positions);
          await message.channel.send(leadingMessage);

          // (Logic bình luận giữ nguyên) ...
          
          raceFinished = raceManager.isRaceFinished(positions, trackLength);
          if (!raceFinished) await new Promise(resolve => setTimeout(resolve, 3000)); 
        }
        
        // --- 4. XỬ LÝ KẾT QUẢ (PHẦN QUAN TRỌNG ĐÃ SỬA) ---
        // Lấy danh sách Top 1 và Top 2
        const { rank1, rank2 } = raceManager.getPodium(positions);
        
        // Tính tiền cược dựa trên Top 1 và Top 2
        const betResults = await betManager.processBetResults(rank1, rank2);
        
        // Tạo chuỗi hiển thị tên ngựa thắng
        const rank1Names = rank1.map(num => `**${raceManager.getHorseName(num)}** (#${num})`).join(', ');
        const rank2Names = rank2.length > 0 
            ? rank2.map(num => `**${raceManager.getHorseName(num)}** (#${num})`).join(', ')
            : "Không xác định";

        const resultEmbed = new EmbedBuilder()
          .setTitle('🏁 KẾT QUẢ CHUNG CUỘC')
          .setColor('#FFD700')
          .setDescription(`
          🥇 **HẠNG 1 (x3.6):** ${rank1Names}
          🥈 **HẠNG 2 (x2):** ${rank2Names}
          `)
          .setTimestamp();
        
        let winnerCount = 0;
        
        // Hiển thị người thắng
        // Gom nhóm kết quả để hiển thị cho gọn nếu 1 người thắng nhiều vé (Optional)
        betResults.forEach(result => {
          if (result.won) {
            winnerCount++;
            let typeText = result.rankType === 'top1' ? '🥇 Top 1' : '🥈 Top 2';
            let emoji = result.rankType === 'top1' ? '🤑' : '😋';
            
            resultEmbed.addFields({
              name: `${emoji} ${typeText} - Ngựa số ${result.horseNumber}`,
              value: `<@${result.userId}> cược ${result.betAmount} ➔ Húp **${result.winAmount} coin**!`,
              inline: false
            });
          }
        });
        
        // Hiển thị người thua All-in
        const allInLosers = betResults.filter(r => !r.won && r.isAllIn);
        if (allInLosers.length > 0) {
            // Lọc trùng ID nếu user đặt nhiều vé all-in (thực tế code chặn 1 lần allin nhưng cứ check cho chắc)
            const uniqueLosers = [...new Set(allInLosers.map(r => r.userId))];
            const loserMentions = uniqueLosers.map(id => `<@${id}>`).join(', ');
            
            resultEmbed.addFields({
                name: '💀 DANH SÁCH RA ĐÊ (Thua All-in)',
                value: `${loserMentions} đã trắng dé và phải ra đê ngủ với dế.`,
                inline: false
            });
        }

        if (winnerCount === 0) {
          resultEmbed.addFields({ name: 'Thua hết!', value: 'Không ai cược đúng. Nhà cái húp trọn!', inline: false });
        }
        
        await message.channel.send({ embeds: [resultEmbed] });
        betManager.clearAllBets();
        
        // Kết thúc
        raceManager.setRaceStatus(false);

    } catch (error) {
        console.error('Lỗi race:', error);
        raceManager.setRaceStatus(false);
        if (fromPrerace) raceManager.setPreraceStatus(false);
        return message.reply('Có lỗi khi đua!');
    }
  },
};

File: cogs/race.py
import asyncio
import discord
from discord.ext import commands
from utils import race_manager, bet_manager
from cogs.ui_components import create_prerace_embed, create_race_start_embed, create_race_result_embed, Colors

class Race(commands.Cog):

    def __init__(self, bot: commands.Bot):
        self.bot = bot
        self.force_start = False

    @commands.command(name='prerace')
    async def prerace(self, ctx: commands.Context):
        if race_manager.is_race_in_progress():
            await ctx.reply('❌ Đang có cuộc đua diễn ra rồi, đợi xong đã!')
            return
        if race_manager.is_prerace_in_progress():
            await ctx.reply('❌ Cổng cược đang mở rồi, nhanh tay đặt cược đi!')
            return
        try:
            race_manager.set_prerace_status(True)
            self.force_start = False
            horses = race_manager.generate_race_names()
            time_left = 60
            total_time = 60
            embed = create_prerace_embed(horses, time_left, total_time)
            msg = await ctx.reply(embed=embed)
            for _ in range(12):
                if self.force_start:
                    await msg.edit(content='🚫 **ĐÃ ĐÓNG CỔNG ỦNG HỘ SỚM!**', embed=None)
                    break
                await asyncio.sleep(5)
                if self.force_start:
                    await msg.edit(content='🚫 **ĐÃ ĐÓNG CỔNG ỦNG HỘ SỚM!**', embed=None)
                    break
                time_left -= 5
                if time_left > 0:
                    updated_embed = create_prerace_embed(horses, time_left, total_time)
                    await msg.edit(embed=updated_embed)
                else:
                    await msg.edit(content='🚫 **ĐÃ HẾT GIỜ ỦNG HỘ!**', embed=None)
                    await ctx.send('🏁 **Hệ thống tự động bắt đầu Chặng đua!**')
                    await self._run_race(ctx, from_prerace=True)
        except Exception as e:
            print(f'Error in prerace: {e}')
            race_manager.set_prerace_status(False)
            await ctx.reply('❌ Đã xảy ra lỗi khi mở cổng cược.')

    @commands.command(name='race')
    async def race_cmd(self, ctx: commands.Context):
        if not race_manager.is_prerace_in_progress():
            await ctx.reply('❌ Chưa mở cổng ủng hộ! Hãy dùng lệnh `!prerace` trước.')
            return
        if race_manager.is_race_in_progress():
            await ctx.reply('❌ Đang có Chặng đua diễn ra rồi, đợi xong đã!')
            return
        self.force_start = True
        await self._run_race(ctx, from_prerace=False)

    async def _run_race(self, ctx: commands.Context, from_prerace: bool=False):
        try:
            bets = bet_manager.get_all_bets()
            if not bets:
                if from_prerace:
                    race_manager.set_prerace_status(False)
                    await ctx.send('❌ Không có ai cược cả, hủy cuộc đua! Trả lại cổng cược.')
                else:
                    await ctx.reply('❌ Không có bet thủ nào tham gia, lấy ai đua?')
                return
            race_manager.set_race_status(True)
            race_manager.set_prerace_status(False)
            start_embed = create_race_start_embed()
            await ctx.send(embed=start_embed)
            await asyncio.sleep(2)
            positions = [0] * race_manager.HORSE_COUNT
            track_length = race_manager.TRACK_LENGTH
            initial_track = race_manager.create_race_status_message(positions, track_length)
            from cogs.ui_components import create_live_race_embed
            status_msg = await ctx.send(embed=create_live_race_embed(initial_track, '*Các mã nương đã sẵn sàng ở vạch xuất phát!*'))
            commentary_history = ['*Các mã nương đã sẵn sàng ở vạch xuất phát!*']
            while not race_manager.is_race_finished(positions, track_length):
                positions = race_manager.simulate_race_step(positions, track_length)
                new_track = race_manager.create_race_status_message(positions, track_length)
                leader_text = race_manager.create_leading_horse_message(positions)
                commentary_history.append(f'*{leader_text}*')
                if len(commentary_history) > 6:
                    commentary_history.pop(0)
                full_commentary = '\n'.join(commentary_history)
                await status_msg.edit(embed=create_live_race_embed(new_track, full_commentary))
                await asyncio.sleep(3)
            podium = race_manager.get_podium(positions)
            rank1 = podium['rank1']
            rank2 = podium['rank2']
            bet_results = await bet_manager.process_bet_results(rank1, rank2)
            rank1_names = ' và '.join([f'**{race_manager.get_horse_name(num)}** (#{num})' for num in rank1])
            rank2_names = ' và '.join([f'**{race_manager.get_horse_name(num)}** (#{num})' for num in rank2]) if rank2 else 'Không có'
            result_embed = create_race_result_embed(rank1_names, rank2_names, bet_results, race_manager.get_horse_name)
            await ctx.send(embed=result_embed)
        except Exception as e:
            print(f'Error in race loop: {e}')
            await ctx.send('❌ Lỗi cmnr hệ thống đền tiền tự chịu trách nhiệm.')
        finally:
            bet_manager.clear_all_bets()
            race_manager.set_race_status(False)
            race_manager.set_prerace_status(False)

async def setup(bot: commands.Bot):
    await bot.add_cog(Race(bot))